Move build instructions from wiki to repository (#193)
aibasel/downward: BUILD.md

Some planner configurations depend on an LP or MIP solver. We support CPLEX (commercial, [free academic license](http://ibm.com/academic)) and SoPlex (Apache License, no MIP support). You can install one or both solvers without causing conflicts.

Once LP solvers are installed and the environment variables `cplex_DIR` and/or `soplex_DIR` are set up correctly, Fast Downward automatically includes each solver detected on the system in the build.

#### Installing CPLEX

Obtain CPLEX and follow the guided installation. See [troubleshooting](#troubleshooting) if you have problems accessing the installer.
On Windows, install CPLEX into a directory without spaces.

After the installation, set the environment variable `cplex_DIR` to the subdirectory `/cplex` of the installation.
For example on Ubuntu:
```bash
export cplex_DIR=/opt/ibm/ILOG/CPLEX_Studio2211/cplex
```
Note that on Windows, setting up the environment variable might require using `/` instead of the more Windows-common `\`.

#### Installing SoPlex on Linux/macOS

**Important:**  The GNU Multiple Precision library (GMP) is critical for the performance of SoPlex but the build does not complain if it is not present.

> Make sure that the build uses the library (check the output of cmake for `Found GMP`).

Reviewer: ```suggestion
Make sure that the build uses the library (check the output of CMake for `Found GMP`).
```

> As of SoPlex 6.0.4, the release does not support C++-20, so we build  from the tip of the [Github main branch](https://github.com/scipopt/soplex) (adapt the path if you install another version or want to use another location):

Reviewer: ```suggestion
As of SoPlex 6.0.4, the release does not support C++-20, so we build  from the tip of the [GitHub main branch](https://github.com/scipopt/soplex) (adapt the path if you install a different version or want to use a different location):
```

```bash
sudo apt install libgmp3-dev
git clone https://github.com/scipopt/soplex.git
export soplex_DIR=/opt/soplex-6.0.4x
export CXXFLAGS="$CXXFLAGS -Wno-use-after-free" # Ignore compiler warnings about use-after-free
cmake -S soplex -B build
cmake --build build
cmake --install build --prefix $soplex_DIR
rm -rf soplex build
```

> After installation, permanently set the environment variable `soplex_DIR` to the value you used in installation.

Reviewer: ```suggestion
After installation, permanently set the environment variable `soplex_DIR` to the value you used during the installation.
```

**Note:** Once [support for C++-20](https://github.com/scipopt/soplex/pull/15) has been included in a SoPlex release, we can update this and can recommend the [SoPlex homepage](https://soplex.zib.de/index.php#download) for downloads instead.

### Optional: Plan Validator

You can validate the found plans by passing `--validate` to the planner if the [VAL plan validation software](https://github.com/KCL-Planning/VAL)

> is installed on your system and the binary `validate` is on the `PATH`.

Reviewer: ```suggestion
is installed on your system and the binary `validate` is on the `PATH`. Note that the `--debug` flag implies `--validate`.
```

**Note:** VAL has a [bug](https://github.com/KCL-Planning/VAL/issues/48) that prevents it from correctly handling the IPC 18 data network domain.

> You can install an older version, e.g. under Debian/Ubuntu:

Reviewer: ```suggestion
You can install an older version, e.g., under Debian/Ubuntu:
```

```bash
sudo apt install g++ make flex bison
git clone https://github.com/KCL-Planning/VAL.git
cd VAL
git checkout a5565396007eee73ac36527fbf904142b3077c74
make clean  # Remove old binaries.
sed -i 's/-Werror //g' Makefile  # Ignore warnings.
make
```

Don't forget to add the resulting `validate` binary to your `PATH`.

## Compiling the planner

To build the planner, from the top-level directory run:

```bash
./build.py
```

This creates the default build `release` in the directory `builds`. For information on alternative builds (e.g. `debug`) and further options, call
`./build.py --help`. [Our website](https://www.fast-downward.org/ForDevelopers/CMake) has details on how to set up development builds.

### Compiling on Windows

Windows does not interpret the shebang in Python files, so you have to call `build.py` as `python3 build.py` (make sure `python3` is on your `PATH`). Also note that options are passed without `--`, e.g., `python3 build.py build=debug`.

> Note that compiling from terminal is only possible with the right environment. The easiest way to get such an environment is to use the `Developer PowerShell for VS 2019` or `Developer PowerShell`.

Reviewer: ```suggestion
Note that compiling from the terminal is only possible with the right environment. The easiest way to get such an environment is to use the `Developer PowerShell for VS 2019` or `Developer PowerShell`.
```
